book/chrome.md

```diff
@@ -715,70 +748,92 @@ class DrawLine:
-    name = "Tab {}".format(i)
-    x1, x2 = 40 + 80 * i, 120 + 80 * i
-```
+Let's reserve some space for all this at the top of the window. But how much?
+One way could be to pick some arbitrary browser chrome height, then try to
+squeeze the chrome into it (and making its font smaller as necessary). Another,
+better way, is to pick a font size that is easy enough to read, then compute
+the chrome height accordingly.
 
-For each tab, we need to create a border on the left and right and
-then draw the tab name:
+Given this design, we can now figure out `self.bottom`: it's the vertical
+height of those two lines plus some padding between and after them. For
+convenience, and to use them later for processing mouse clicks, I'll store
+these parameters on the `Browser` object:^[I also chose `20px` as the
+font size. Depending on your computer, this may end up looking smaller,
+\index{device pixel ratio}
+because of the device pixel ratio of the screen.]
 
 ``` {.python}
-for i, tab in enumerate(self.tabs):
-    # ...
-    cmds.append(DrawLine(x1, 0, x1, 40, "black", 1))
-    cmds.append(DrawLine(x2, 0, x2, 40, "black", 1))
-    cmds.append(DrawText(x1 + 10, 10, name, tabfont, "black"))
+class Chrome:
+    def __init__(self, browser):
+        # ...
+        self.font = get_font(20, "normal", "roman")
+        font_height = self.font.metrics("linespace")
+
+        self.padding = 5
+        self.tab_header_bottom = font_height + 2 * self.padding
+        self.addressbar_top = self.tab_header_bottom + self.padding
+        self.bottom = \
+            self.addressbar_top + font_height + \
+            2 * self.padding    
 ```
 
-Finally, to identify which tab is the active tab, we've got to make
-that file folder shape with the current tab sticking up:
+In addition, it's convenient to define some methods that return the bounds of
+the plus and each tab. Note how the size of each
+elements' bounds is as wide as the font says it is. That way, if we change our
+mind later, we can just change the font size, and everything else just draws
+correctly.^[Also, we chose "Tab 1" as the fixed size of a tab, but in real
+life many fonts draw different numbers (like "1" and "6") with different
+widths, and this approximation also doesn't work for two-digit tab numbers.
+Real browsers size their tabs to fit the [title], but we skipped that for
+simplicity.]
+
+[title]: https://developer.mozilla.org/en-US/docs/Web/API/Document/title
 
 ``` {.python}
-for i, tab in enumerate(self.tabs):
-    # ...
-    if i == self.active_tab:
-        cmds.append(DrawLine(0, 40, x1, 40, "black", 1))
-        cmds.append(DrawLine(x2, 40, WIDTH, 40, "black", 1))
+class Chrome:
+    def plus_bounds(self):
+        plus_width = self.font.measure("+")
+        return (self.padding, self.padding,
+            plus_width + self.padding,
+            self.tab_header_bottom - self.padding)
+
+    def tab_bounds(self, i):
+        tab_start_x = self.font.measure("+") + \
```

Comment:
Some of these `bounds` methods I'd like to be a little nicer to read. Some options:

- Minimally, can we do `self.padding + self.font.measure("+") + self.padding`, reflecting the actual left-to-right order?
- Or can we call `self.plus_bounds` and do offsets from there?
- I still think we could do better if it was one `layout` method, but I agree we shouldn't delay this PR